hathor/indexes/rocksdb_deps_index.py

```diff
@@ -0,0 +1,329 @@
+            batch.delete((self._cf, self._to_key_reverse(tx_dep_hash, tx.hash)))
+
+    def _add_ready(self, tx_hash: bytes, batch: 'rocksdb.WriteBatch') -> None:
+        key = self._to_key_ready(tx_hash)
+        batch.put((self._cf, key), b'')
+
+    def add_tx(self, tx: BaseTransaction, partial: bool = True) -> None:
+        import rocksdb
+        assert tx.hash is not None
+        assert tx.storage is not None
+        batch = rocksdb.WriteBatch()
+        validation = tx.get_metadata().validation
+        if validation.is_fully_connected():
+            self._del_from_deps(tx, batch)
+        elif not partial:
+            raise ValueError('partial=False will only accept fully connected transactions')
+        else:
+            self._add_deps(tx, batch)
+            self._add_needed(tx, batch)
+        self._db.write(batch)
+
+    def del_tx(self, tx: BaseTransaction) -> None:
+        import rocksdb
+        batch = rocksdb.WriteBatch()
+        self._del_from_deps(tx, batch)
+        self._db.write(batch)
+
+    def _add_deps(self, tx: BaseTransaction, batch: 'rocksdb.WriteBatch') -> None:
+        assert tx.hash is not None
+        for dep in tx.get_all_dependencies():
+            batch.put((self._cf, self._to_key_reverse(dep, tx.hash)), b'')
+
+    def _add_needed(self, tx: BaseTransaction, batch: 'rocksdb.WriteBatch') -> None:
+        assert tx.hash is not None
+        assert tx.storage is not None
+        tx_storage = tx.storage
+
+        height = get_requested_from_height(tx)
+        self.log.debug('add needed deps', tx=tx.hash_hex, height=height, type=type(tx).__name__)
+        # get_tx_parents is used instead of get_tx_dependencies because the remote will traverse the parent tree, not
+        # the dependency tree, eventually we should receive all tx dependencies and be able to validate this tx
+        for tx_dep_hash in tx.get_tx_parents():
+            # It may happen that we have one of the dependencies already, so just add the ones we don't have. We should
+            # add at least one dependency, otherwise this tx should be full validated
+            if not tx_storage.transaction_exists(tx_dep_hash):
+                self.log.debug('tx parent is needed', tx=tx.hash.hex(), tx_dep=tx_dep_hash.hex())
+                batch.put((self._cf, self._to_key_needed(tx_dep_hash)), self._to_value_needed(height, tx.hash))
+
+    def remove_ready_for_validation(self, tx: bytes) -> None:
+        self._db.delete((self._cf, self._to_key_ready(tx)))
+
+    def next_ready_for_validation(self, tx_storage: 'TransactionStorage', *, dry_run: bool = False) -> Iterator[bytes]:
+        import rocksdb
+        batch = rocksdb.WriteBatch()
+        ready = self._drain_all_sorted_ready(tx_storage, batch)
+        if not dry_run:
+            self._db.write(batch)
+        while ready:
+            yield from ready
+            batch = rocksdb.WriteBatch()
+            ready = self._drain_all_sorted_ready(tx_storage, batch)
```

Comment:
What happens if the full node is stopped after the generator has started, all items have been drained, but the caller hasn't finished consuming the generator? Would it recover nicely after the full node is restarted?

Reply:
Yeah, that might be a problem, I'm not sure if it would recover nicely, and even if moving the `_db.write(batch)` to _after_ the `yield from`, so maybe batching them isn't good for recovery, I also found other recovery problems when testing something else with rocksdb indexes. It might not be so trivial to work around these. Maybe switching back to a memory-only index for these one could be the short-term solution.

Reply:
Which other problems have you found? In which indexes?

I'm ok to go ahead and merge this PR without this fix. But I feel it is a mandatory fix before we start using the RocksDBDepsIndex class in production. Maybe we should raise an exception on `__init__()` just to make sure we don't use it by mistake.

Reply:
I don't know if it's a problem on other indexes in particular, but when testing with rocksdb indexes and exiting correctly (SIGINT/Ctrl+C) _while it was syncing_, I got a transaction not found at the end of the loading phase because the transaction was added to the indexes/metadata used for the loading (happens when using either the children metadata or the timestamp index, which are the two "fast iterators" that we have). So, there's some path (and it might be very common) that when interrupted, it will not have saved a new transaction, but will have added it to the indexes.

Prohibiting `__init__()` of the class isn't great because we won't be able to test it. Maybe prefixing the name with something like `DoNotUse` is enough, I think it just has to be evident enough to not be missed in a review.

Reply:
Got it. It's important to keep consistency when gracefully stopping the full node. Should we keep a separate issue to investigate it?

Reply:
I think it makes sense. Created #411

Reply:
I feel that we should add another method to mark the tx validation as completed (instead of draining the records from disk, moving them to memory and taking the risk of losing it all if the daemon is terminated). What do you think?

Reply:
That would work too. I'm to sure that by itself would fix the problem, but it could simplify the implementation of `next_ready_for_validation`. What I imagine is that `remove_ready_for_validation` would become something like `mark_as_complete`, which itself will trickle the updating of the state of the dependencies to check which have become ready to be validated. But we still need a mechanism to not update the list that is currently being iterated through, which I think is the main point that causes issues. The current solution copies the list to prevent modifying it while iterating over it. Whether we calculate when a transaction can be validated inside _or_ outside of this function, it doesn't change the fact that we still have a common "logical list" that we're adding to and removing from in the same loop.

Reply:
I think we can add a note about it with the link to the issue and go ahead. This index is not in use right now, and we will go through it on the sync-v2 PR.

Reply:
Great! Thanks!